### PR TITLE
reimplement parsing

### PR DESCRIPTION
**Pull request by connorskees**
This pull request involves a complete rewrite of parsing, allowing us to simplify parsing as well as `content-exists` and `@import` in all contexts. This also involves a complete rewrite of control flow evaluation, resulting in something much more robust. Scoping of variables is also much more accurate.

There should not be any performance regressions and all tests are passing currently. 

#### What's next
 - `@extend`, which should be much easier as a result of this refactor
 - `@keyframes`
 - `@supports`

### CODE REVIEW

File: src/parse/common.rs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+            first,
+            rest: Vec::new(),
+        }
+    }
+
+    pub fn last(&self) -> &T {
+        match self.rest.last() {
+            Some(v) => v,
+            None => &self.first,
+        }
```
Comment (pickfire) [inline src/parse/common.rs:24]:
```suggestion
        self.rest.last().unwrap_or(&self.first)
```
Not tested but probably works, same goes for the one below.

File: src/parse/mod.rs [hunk 2]
```diff
@@ -0,0 +1,1937 @@
...
+        p1.push(OsString::from("index.scss"));
+        let mut p2 = path_buf.clone();
+        p2.push(OsString::from("_index.scss"));
+        let paths = [
+            path_buf.with_file_name(name).with_extension("scss"),
+            path_buf.with_file_name(format!("_{}.scss", name.to_str().unwrap())),
+            path_buf,
+            p1,
+            p2,
+        ];
```
Comment (pickfire) [inline src/parse/mod.rs:1745]: Why not? (not tested)
```suggestion
        let paths = [
            path_buf.with_file_name(name).with_extension("scss"),
            path_buf.with_file_name(format!("_{}.scss", name.to_str().unwrap())),
            path_buf,
            path_buf.join("index.scss"),
            path_buf.join("_index.scss"),
        ];
```

File: src/parse/mod.rs [hunk 3]
```diff
@@ -0,0 +1,1937 @@
...
+        let path: &Path = file_name.as_ref();
+
+        let mut rules = Vec::new();
+        let path_buf = if path.is_absolute() {
+            // todo: test for absolute path imports
+            path.into()
+        } else {
+            self.path
+                .parent()
+                .unwrap_or_else(|| Path::new(""))
```
Comment (pickfire) [inline src/parse/mod.rs:1726]:
```suggestion
                 .unwrap_or_default()
```

File: src/parse/mod.rs [hunk 4]
```diff
@@ -0,0 +1,1937 @@
...
+            p1,
+            p2,
+        ];
+        for name in &paths {
+            if name.is_file() {
+                let file = self.map.add_file(
+                    name.to_string_lossy().into(),
+                    String::from_utf8(fs::read(name)?)?,
+                );
+                let rules2 = Parser {
```
Comment (pickfire) [inline src/parse/mod.rs:1752]: 2?

File: src/parse/mod.rs (hunk 2 above)
Comment (connorskees) [reply src/parse/mod.rs:1745]: Good point!

File: src/parse/mod.rs (hunk 3 above)
Comment (connorskees) [reply src/parse/mod.rs:1726]: Unfortunately `Path` does not implement `Default`, so this change cannot be made. I wonder if there's a reason it doesn't..

File: src/parse/value.rs [hunk 5]
```diff
@@ -0,0 +1,1121 @@
...
+                    red,
+                    green,
+                    blue,
+                    alpha,
+                    format!("#{}", s),
+                )))
+                .span(self.span_before))
+            }
+            _ => Err(("Expected hex digit.", self.span_before).into()),
+        }
```
Comment (pickfire) [inline src/parse/value.rs:629]: Uses `u32` integer for different types but I don't think it affects the performance.
```suggestion
        let v = match u32::from_str_radix(&s, 16) {
            Ok(a) => a,
            Err(_) => {
                return Ok(Value::String(format!("#{}", s), QuoteKind::None).span(self.span_before))
            }
        };
        let (red, green, blue, alpha) = match s.len() {
            3 => (
                (((v & 0x0f00) >> 8) * 0x11) as u8,
                (((v & 0x00f0) >> 4) * 0x11) as u8,
                ((v & 0x000f) * 0x11) as u8,
                1,
            ),
            4 => (
                (((v & 0xf000) >> 12) * 0x11) as u8,
                (((v & 0x0f00) >> 8) * 0x11) as u8,
                (((v & 0x00f0) >> 4) * 0x11) as u8,
                ((v & 0x000f) * 0x11) as u8,
            ),
            6 => (
                ((v & 0x00ff_0000) >> 16) as u8,
                ((v & 0x0000_ff00) >> 8) as u8,
                (v & 0x0000_00ff) as u8,
                1,
            ),
            8 => (
                ((v & 0xff00_0000) >> 24) as u8,
                ((v & 0x00ff_0000) >> 16) as u8,
                ((v & 0x0000_ff00) >> 8) as u8,
                (v & 0x0000_00ff) as u8,
            ),
            _ => return Err(("Expected hex digit.", self.span_before).into()),
        };
        let color = Color::new(red, green, blue, alpha, format!("#{}", s));
        Ok(Value::Color(Box::new(color)).span(self.span_before))
```

File: src/parse/mod.rs (hunk 3 above)
Comment (pickfire) [reply src/parse/mod.rs:1726]: Oh, I didn't know such things exist, maybe people might have different opinion on default for path which is `.`.

File: src/parse/common.rs (hunk 1 above)
Comment (connorskees) [reply src/parse/common.rs:24]: Thanks! Resolved in 822b9a35fefb8726814abda72d6b9f693b9f5352

File: src/parse/mod.rs (hunk 4 above)
Comment (connorskees) [reply src/parse/mod.rs:1752]: Resolved in 191f2dfba4ac401c660f54383298f7dd848c877f

**Pull request merged**